File: compression/techniques/python/rle.py
# Isaac Computer Science
# Usage licensed under the Open Government Licence v3.0
# https://isaaccomputerscience.org/concepts/prog_sub_using_subroutines

def rle(text_string):
    """A simple version of RLE for text strings"""
    current_token = text_string[0]
    compressed = current_token
    counter = 0
    for i in range (len(text_string)):
        next_token = text_string[i]
        if next_token != current_token:
            compressed = compressed + str(counter) + next_token
            counter = 1
        else:
            counter = counter + 1
        current_token = next_token
    compressed = compressed + str(counter)
    return compressed


def run_tests():
    """sample test data"""
    # Normal data
    text = "aaaabbbccdddeee"
    print(text)
    print(rle(text))
    text = "Bobby"
    print(text)
    print(rle(text))
    # Boundary data
    text = "a"
    print(text)
    print(rle(text))
    text = "aaaaaa"
    print(text)
    print(rle(text))    
    text = "aaaae"
    print(text)
    print(rle(text))
    text = "aeeeee"
    print(text)
    print(rle(text))
    
    
# This code will run if this file is executed directly (i.e. not called by another program)
if __name__ == '__main__':
    run_tests()

File: compression/techniques/c-sharp/rle.cs
/*
Isaac Computer Science
Usage licensed under the Open Government Licence v3.0
Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{

    class CompressionTechniques
    {

        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            // Test data - normal data
            string text = "aaaabbbccdddeee";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));

            text = "Bobby";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));

            // Test data - boundary data
            text = "a";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));

            text = "aaaaaa";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));

            text = "aaaae";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));

            text = "aeeeee";
            Console.WriteLine(text);
            Console.WriteLine(RLE(text));
        }

        // A simple version of Run Length Encoding for text strings
        public static string RLE(string textString)
        {
            char currentToken = textString[0];
            string compressed = currentToken.ToString();
            int counter = 1;

            for (int i = 1; i < textString.Length; i++) {
                char nextToken = textString[i];
                if (nextToken != currentToken) {
                    compressed = compressed + counter.ToString() + nextToken;
                    counter = 1;
                }
                else {
                    counter = counter + 1;
                }
                currentToken = nextToken;
            }
            compressed = compressed + counter.ToString();
            return compressed;
        }
    }

}